File: python/quick_sort.py
import time
from constants import *
import sys

sys.setrecursionlimit(30000)


def partition(arr, low, high):
    i = (low - 1)
    pivot = arr[high]

    for j in range(low, high):
        if arr[j] < pivot:
            i = i + 1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1


def quick_sort(arr, low, high):
    if low < high:
        pivot_index = partition(arr, low, high)
        quick_sort(arr, low, pivot_index - 1)
        quick_sort(arr, pivot_index + 1, high)


if __name__ == '__main__':
    for z in range(50):
        set_array = set2()
        n = len(set_array)
        start = time.time() * 1000
        quick_sort(set_array, 0, n - 1)
        end = time.time() * 1000
        elapsed = end - start
        print(elapsed)

File: rust/src/quick_sort.rs
// Code originally forked from https://github.com/diptangsu/Sorting-Algorithms/
include!("constants.rs");
use std::time::Instant;

fn get_partition<T: Ord>(array: &mut [T], low: isize, high: isize) -> isize {
    let pivot_index = high as usize;
    let mut save_index = low - 1;
    let mut last_index = high;

    loop {
        save_index += 1;
        while array[save_index as usize] < array[pivot_index] {
            save_index += 1;
        }
        last_index -= 1;
        while last_index >= 0 && array[last_index as usize] > array[pivot_index] {
            last_index -= 1;
        }
        if save_index >= last_index {
            break;
        } else {
            array.swap(save_index as usize, last_index as usize);
        }
    }
    array.swap(save_index as usize, pivot_index as usize);
    save_index
}

pub fn quick_sort<T: Ord>(array: &mut [T]) {
    let length = array.len();
    _quick_sort(array, 0, (length - 1) as isize);
}

fn _quick_sort<T: Ord>(array: &mut [T], low: isize, high: isize) {
    if low < high {
        let p = get_partition(array, low, high);
        _quick_sort(array, low, p - 1);
        _quick_sort(array, p + 1, high);
    }
}

fn main() {
    unsafe {
        for _x in 0..50 {
            let start = Instant::now();
            quick_sort(&mut SET2);
            let elapsed = start.elapsed();
            println!("{:?}", elapsed);
        }
    }
}